apps/shell/tash_command.c

```diff
@@ -175,6 +194,150 @@ static int tash_exit(int argc, char **args)
 	exit(0);
 }
 
+/** @brief Help function in TASH to list all available commands
+ *  @ingroup tash
+ */
+
+static int tash_history(int argc, char **args)
+{
+	int cmd_idx = 1;
+
+	printf("\t TASH command history\n");
+	printf("\t --------------------\n");
+
+	int head_idx = cmd_head;
+	while (head_idx != cmd_tail) {
+		printf(" %d \t %s\n", cmd_idx++, cmd_store[head_idx]);
+		if (head_idx == TASH_MAX_STORE - 1) {
+			head_idx = 0;
+		} else {
+			head_idx++;
+		}
+	}
+
+	return 0;
+}
+
+void tash_get_history(int num, char *cmd)
+{
+	int cmd_idx = 1;
+
+	if (num < 1 || num >= TASH_MAX_STORE) {
+		return;
+	}
+
+	int head_idx = cmd_head;
+	int pos = 0;
+	while (head_idx != cmd_tail) {
+		if (num == cmd_idx) {
+			while (cmd_store[head_idx][pos] != 0) {
+				cmd[pos] = cmd_store[head_idx][pos];
+				pos++;
+			}
+		}
+		cmd_idx++;
+		if (head_idx - 1 == TASH_MAX_STORE) {
+			head_idx = 0;
+		} else {
+			head_idx++;
+		}
+	}
+}
+
+void tash_store_cmd(char *cmd)
+{
+	/* If there is no command, it is not saved. */
+	if (cmd == NULL || cmd[0] == '\0') {
+		return;
+	}
+
+	if (cmd_head != cmd_tail) {
+		/* If it is the same as the previous command, it is not saved. */
+		int prev;
+		if (cmd_tail == 0) {
+			prev = TASH_MAX_STORE - 1;
+		} else {
+			prev = cmd_tail - 1;
+		}
+
+		if (strncmp(cmd_store[prev], cmd, TASH_LINEBUFLEN) == 0) {
+			return;
+		}
+	}
+
+	/* Save current command. */
+	strncpy(cmd_store[cmd_tail], cmd, TASH_LINEBUFLEN);
+	if (cmd_tail == TASH_MAX_STORE - 1) {
+		cmd_tail = 0;
+	} else {
+		cmd_tail++;
+	}
+
+	/* Move the head when the storage space is full. */
+	if (cmd_tail == cmd_head) {
+		if (cmd_head == TASH_MAX_STORE - 1) {
+			cmd_head = 0;
+		} else {
+			cmd_head++;
+		}
```

Comment:
How about using some definition to avoid same codes?
```
#define CMD_INDEX_UP(x)                                   \
	do {                                                  \
		if ((x) == TASH_MAX_STORE - 1) ? (x) = 0 : (x)++; \
	while (0)

#define CMD_FIND_PREVCMD(x)                                   \
	do {                                                  \
		if (cmd_tail == 0) ? (x) = TASH_MAX_STORE - 1 : (x) = cmd_tail - 1; \
	while (0)

void tash_store_cmd(char *cmd)
{
	(snip)

	/* Save current command. */
	strncpy(cmd_store[cmd_tail], cmd, TASH_LINEBUFLEN);
	CMD_INDEX_UP(cmd_tail);

	/* Move the head when the storage space is full. */
	if (cmd_tail == cmd_head) {
		CMD_INDEX_UP(cmd_head);
	}
	cmd_pos = cmd_tail;
```